### ISSUE TITLE
Prometheus high memory usage(>960Gib) with 1200+ ServiceMonitors

### ISSUE THREAD

**Issue by mramesh047** [prometheus/prometheus#13584]
### What did you do?

We have been running a prometheus on one of our large production clusters which emits >5M+ active time series, which is operator managed Prometheus running on 1024Gib machine. However, prometheus keep crashing with OOM very frequently 

There are around 1200+ serviceMonitors running with 5mins interval, not sure if this something causing memory leak ?

here is pprof top output:
<img width="858" alt="image" src="https://github.com/prometheus/prometheus/assets/139512081/6f2c56cb-5782-4b7a-9c63-31c18988bb2c">

<img width="438" alt="Screenshot 2024-02-14 at 6 57 19 PM" src="https://github.com/prometheus/prometheus/assets/139512081/0fc4c98f-606a-47ef-874c-613b3a520adc">

Sample labeling for one of the targets:(95% of targets have same set of labels)
<img width="485" alt="Screenshot 2024-02-14 at 9 45 21 PM" src="https://github.com/prometheus/prometheus/assets/139512081/043a542c-f366-4ba1-96bd-89c32d38a85c">

### What did you expect to see?

I feel 1024Gib is quite large to manage 50M+ timeseries

### What did you see instead? Under which circumstances?

Memory usage getting spiked to 900Gib+ with in 10-15mins which is unusual

<img width="1404" alt="Screenshot 2024-02-14 at 10 00 04 PM" src="https://github.com/prometheus/prometheus/assets/139512081/35e1b8df-b2a8-4ab3-ad97-346d2b00520a">

Memory usage based on `container_memory_working_set_bytes` metric
<img width="1515" alt="image" src="https://github.com/prometheus/prometheus/assets/139512081/6f330e94-a4b4-48e0-84b1-0af70a2720ae">

### Prometheus version

```text
v2.45+stringlabels
```

**Comment by mramesh047** [comment]
Thanks @bboreham for your response. yes we will be in process of upgrading to v2.49 in next few weeks.

I am still curious as to why it's eating so much RAM for the limited series and targets. Is there anything that we need to optimize from our end like grouping more endpoints with minimal number of serviceMonitors?


Here is the pprof file
[heap_wd101.zip](https://github.com/prometheus/prometheus/files/14297964/heap_wd101.zip)

**Comment by bboreham** [comment]
> Prometheus version v2.45+stringlabels

This is not a tag we published on Dockerhub, and your profile does not match 2.45.
It is a much better match to 2.43.

From the profile, it looks like your service discovery is collecting 35 million targets.
I don't use ServiceMonitor myself, so can't advise on the details of its usage, but certainly try to trim that number.

Viewing `inuse_objects`:
<img width="806" alt="image" src="https://github.com/prometheus/prometheus/assets/8125524/72c39ac2-0ed5-44e3-a900-39147d300329">

**Comment by mramesh047** [comment]
My bad it was 2.43+string labels

Thanks @bboreham for your response and providing more details

**Comment by machine424** [comment]
The `targets` leak may be related to https://github.com/prometheus/prometheus/issues/12858 for which I'm proposing a fix here: https://github.com/prometheus/prometheus/pull/13147

Or related to this other leak here: https://github.com/prometheus/prometheus/issues/12436 which I think we can address once https://github.com/prometheus/prometheus/pull/13147 is merged as the solution I'm thinking for relies on it.

In the meantime I'll raise a PR for https://github.com/prometheus/prometheus/issues/12436 as well.

**Comment by bboreham** [comment]
I note that #13147 changes only the experimental service discovery manager, not the one used here.

**Comment by machine424** [comment]
> I note that https://github.com/prometheus/prometheus/pull/13147 changes only the experimental service discovery manager, not the one used here.

You're right, and I'm only planning to fix https://github.com/prometheus/prometheus/issues/12436 for the  "experimental service discovery manager". I think it's time that the new manager takes over, besides these fixes, there are a couple of other bugs that only occur on the legacy manager.

I'll create an issue for the switch to discuss that.

**Comment by mramesh047** [comment]
<img width="758" alt="image" src="https://github.com/prometheus/prometheus/assets/139512081/3f8e7a5d-c8d0-4f8a-991f-efa6bf201949">

There is a massive improvement in memory usage after optimizing the serviceMonitors from 1:1 mapping to 1:mulitple

**Comment by bboreham** [comment]
I can't tell much from that information; if you upload another heap pprof it may give some clues.

The way I use the term, a "leak" would show an increase over time, or at least a drop on restart of the process doing the same work.  So there is no leak in the picture, for me.

**Comment by mramesh047** [comment]
[wd101-heap-with-improvements.zip](https://github.com/prometheus/prometheus/files/14391819/wd101-heap-with-improvements.zip)

This is the one got collected when kc top commands as below

```
➜  kc top po -n application-monitoring
NAME                                 CPU(cores)   MEMORY(bytes)
prometheus-0          162m         477217Mi
```

**Comment by bboreham** [comment]
That profile shows 43GiB allocated while rendering a web page showing dropped targets.
There have been a couple of improvements since v2.43 that should help:
* v2.47 added `keep_dropped_targets` which can limit the amount of detail retained in memory (e.g. set it to 10)
* v2.50 avoids an intermediate map. #13484

Also you could look at narrowing the selection of targets, e.g. using selectors, so you don't have to drop so many.

```
github.com/prometheus/prometheus/web/api/v1.(*API).targets
/go/src/github.com/prometheus/prometheus/web/api/v1/api.go

  Total:      6.59MB    43.93GB (flat, cum) 66.55%
   1010            .          .           	if showDropped { 
   1011            .          .           		targetsDropped := api.targetRetriever(r.Context()).TargetsDropped() 
   1012            .          .           		droppedKeys, numTargets := sortKeys(targetsDropped) 
   1013       4.09MB     4.09MB           		res.DroppedTargets = make([]*DroppedTarget, 0, numTargets) 
   1014            .          .           		for _, key := range droppedKeys { 
   1015            .          .           			if scrapePool != "" && key != scrapePool { 
   1016            .          .           				continue 
   1017            .          .           			} 
   1018            .          .           			for _, target := range targetsDropped[key] { 
   1019       2.50MB     2.50MB           				res.DroppedTargets = append(res.DroppedTargets, &DroppedTarget{ 
   1020            .    43.92GB           					DiscoveredLabels: target.DiscoveredLabels().Map(), 
   1021            .          .           				}) 
   1022            .          .           			} 
   1023            .          .           		} 
```

**Comment by leegin** [comment]
@mramesh047 We have observed high CPU usage when we had one service monitor per service. We didn't see any spike with memory because of this. We also went with 1: many mapping to solve the issue.

**Comment by SuperQ** [comment]
Note `container_memory_working_set_bytes` includes some kernel cache metrics. You want to look at `container_memory_rss` to find out how much Prometheus itself is using.

**Comment by bboreham** [comment]
@mramesh047 did you try any of the suggestions?  Did things get any better?

**Comment by bboreham** [comment]
Hello from the bug-scrub!  

One more improvement was made - #15261 in 3.2.0, also aimed at reducing memory usage with a lot of ServiceMonitors.
Given no response in 5 months we will close this issue.

**Comment by ringerc** [comment]
@SuperQ To get a realistic idea of Prometheus's true memory use you also need to set an aggressive `GOGC`.

And/or need to set set a `GOMEMLIMIT` or corresponding auto-gomemlimit ratio that puts the GC threshold close to the program's current `go_memstats_heap_inuse_bytes + go_memstats_mcache_inuse_bytes + go_memstats_mspan_inuse_bytes + go_memstats_stack_inuse_bytes` (though the latter 3 are usually small enough to be irrelevant. But not too close, otherwise the go runtime will thrash too much on CPU trying to work within a too-small memory footprint.

Otherwise the go runtime won't bother to find or clean up memory it has allocated, and lots of the seemingly "in use" memory will just be Go's undifferentiated garbage heap of abandoned-but-not-yet-reclaimed objects. These won't necessarily even show up in `go_memstats_heap_idle_bytes` if the program isn't under memory pressure so the GC is running very lazily.

Using `container_memory_rss` is a lot better than the working set for true Prometheus memory use, but it's still counting a bunch of go garbage collection noise.

**Comment by SuperQ** [comment]
Yes, I'm very aware of Go memory metrics. The reason I recommend `container_memory_rss` is because this is what the OS sees. It is what the Go runtime has allocated _in the OS_. Other Go memory metrics are useful for optimizing GC overhead, but not for monitoring at the OS level.
